### PR TITLE
added aci_interface_policy_leaf_policy_group module

### PR DESCRIPTION
**Pull request by brunocalogero**
##### SUMMARY
Adding aci_interface_policy_leaf_policy_group module, creates the following ACI objects: infra:AccPortGrp and infra:AccBndlGrp

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aci_interface_policy_leaf_policy_group

##### ANSIBLE VERSION
```
2.5
```

### CODE REVIEW

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+'''
+
+RETURN = ''' # '''
+
+from ansible.module_utils.network.aci.aci import ACIModule, aci_argument_spec
+from ansible.module_utils.basic import AnsibleModule
+
+
+def main():
+    argument_spec = aci_argument_spec
```
Comment (sivel) [inline lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: `aci_argument_spec` has been updated to be a function.  This line should be updated as:

```
argument_spec = aci_argument_spec()
```

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py (hunk 1 above)
Comment (brunocalogero) [reply lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: alrighty ! 👍

**Comment by brunocalogero** [comment]
`2018-01-17 21:33:18 ERROR: Found 14 validate-modules issue(s) which need to be resolved:
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "cert_key" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "cert_name" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "certificate_name" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "host" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "hostname" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "password" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "private_key" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "protocol" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "timeout" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "use_proxy" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "use_ssl" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "user" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "username" is listed in the argument_spec, but not documented in the module (75%)
2018-01-17 21:33:18 ERROR: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py:0:0: E322 "validate_certs" is listed in the argument_spec, but not documented in the module (75%)`, are there new documentation rules ?

**Comment by sivel** [comment]
> are there new documentation rules

Yes, the new rules were activated today after testing.

The errors indicate inconsistencies between the `DOCUMENTATION` and the `argument_spec`. Looking at those errors, it may indicate at a minimum that you are missing a `extends_documentation_fragment` in your `DOCUMENTATION`.

The bot should likely also be updating this PR with a comment soon in regards to those errors.

**Comment by brunocalogero** [comment]
@sivel hey what do you suggest to fix the protocol documentation error ? what are the possible options for protocol, strange that adding `extends_documentation_fragment: aci` fixed everything except this one

**Comment by sivel** [comment]
@dagwieers is `protocol` one of the arguments that is purposefully not documented?

**Comment by dagwieers** [comment]
@sivel Yes, it's replaced with `use_ssl` (in line with other modules). It's now fixed as we moved `protocol` out of the standard `aci_argument_spec` since it's being deprecated and only used for v2.4 modules, not the new stuff.

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+   - Choice of l2_interface_policy to be used as part of the leaf policy group to be created.
+   aliases: [ l2_interface_policy_name ]
+ port_security_policy:
+   description:
+   - Choice of port_security_policy to be used as part of the leaf policy group to be created.
+   aliases: [ port_security_policy_name ]
+ attached_entity_profile:
+   description:
+   - Choice of attached_entity_profile (AEP) to be used as part of the leaf policy group to be created.
+   aliases: [ attached_entity_profile_name, AEP, AEP_name ]
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: We use `aep` as the default parameter in other modules and we never used `AEP`, `AEP_name` or `attached_entity_profile`. So please use the same parameter name as the **aci_aep** module !

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+   - Name of the leaf policy group to be added/deleted.
+   aliases: [ name, policy_group_name ]
+ description:
+   description:
+   - Description for the leaf policy group to be created.
+   aliases: [ descr, description ]
+ link_aggregation_type:
+   description:
+   - Selector for the type of leaf policy group we want to create.
+   aliases: [ link_aggregation_type_name ]
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: I wonder if this wouldn't be better named `lag_type` either as the default name, or an alias. As LAG seems to be a common term (at least I know this term).
cc @rsmeyers @jmcgill298

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+short_description: Add Fabric Interface Policy Leaf Policy Groups on Cisco ACI fabrics.
+description:
+- Add Fabric Interface Policy Leaf Policy Groups on Cisco ACI fabrics.
+- More information from the internal APIC class I(infra:AccBndlGrp), I(infra:AccPortGrp) at
+  U(https://developer.cisco.com/site/aci/docs/apis/apic-mim-ref/).
+author:
+- Bruno Calogero (@brunocalogero)
+version_added: '2.5'
+notes:
+- When using the module please select the appropriate link_aggregation_type. 'link'(PC)/'node'(VPC), 'leaf'(Leaf Access Port Policy Group)
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: You should be doing: `C(link) for PC, C(node) for VPC and C(leaf) for Leaf Access Port Policy Group`.

Maybe also use the full name for PC and VPC as well ?

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+
+from ansible.module_utils.network.aci.aci import ACIModule, aci_argument_spec
+from ansible.module_utils.basic import AnsibleModule
+
+
+def main():
+    argument_spec = aci_argument_spec()
+    argument_spec.update({
+        'policy_group': dict(type='str', aliases=['name', 'policy_group_name']),
+        'description': dict(type='str', aliases=['descr', 'description']),
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: Please remove `description' from the aliases.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py (hunk 3 above)
Comment (brunocalogero) [reply lib/ansible/modules/network/aci/aci_interface_policy_leaf_policy_group.py]: agreed, was just wondering if we should make it same as GUI or not 👍

**Review by dagwieers — APPROVED** [review]
LGTM

**Comment by brunocalogero** [comment]
@sivel can you approve changes ?

**Review by sivel — APPROVED** [review]

**Comment by sivel** [comment]
I've approved the changes I asked for.  I'll leave this in @dagwieers' hands, as I have no expertise in this specific area (aci).

**Comment by dagwieers** [comment]
I made one more commit to it, but lost track waiting for it to be validated.

**Comment by dagwieers** [comment]
PS Please don't forget to add the integration test in due time.

**Pull request merged**